src/core/midiPorts.cpp

```diff
@@ -0,0 +1,426 @@
+			return 2;
+		}
+		// However if it is in the map but not open, something is wrong
+		// Oh well, we'll se what happens next.
+	}
+
+	// No entry in map, so we shall create one
+	else {
+		try {
+			inPorts_[port] = new RtMidiIn((RtMidi::Api) api_,
+						"Giada Input to " + port);
+		}
+		catch (RtMidiError& error) {
+			u::log::print("[MP::openInPort] RtMidiError: %s\n",
+						error.getMessage().c_str());
+			return -2;
+		}
+	}
+	
+	// Let's open this port
+	try {
+		// Create a persistent string with a port name, so callback
+		// function knows which port did the data come from. 
+		// TODO: How do we dispose of this thing when the port is closed?
+		// A unique_ptr would be nice but I think it's not applicable here.
+		std::string* p = new std::string(port);
+
+		inPorts_[port]->setCallback(&callback_, p);
+		inPorts_[port]->openPort(index, port);
+		inPorts_[port]->ignoreTypes(false, false, false);
+		u::log::print("[MP::openInPort] MIDI port \"%s\" open\n",
+							port.c_str());
+		return 1;
+	}
+	catch (RtMidiError& error) {
+		u::log::print("[MP::openInPort] unable to open port %s: %s\n",
+				port.c_str(), error.getMessage().c_str());
+		return -3;
+	}
+
+}
+
+/* -------------------------------------------------------------------------- */
+
+int closeOutPort(std::string port){
+
+	if (port == ""){
+
+		// Create a list of open ports and destroy them all!
+		std::vector<std::string> ports_to_destroy;
+
+		for (auto const& [name, rtm] : outPorts_){
+			ports_to_destroy.push_back(name);
+		}
+
+		for (unsigned int i = 0; i < ports_to_destroy.size(); i++){
+			closeOutPort(ports_to_destroy.at(i));
+		}
+		return 1;
+	}
+	// Let's check if this port is in the map //
+	if (outPorts_.count(port) > 0) {
+		// RtMidi port destructor closes connections on its own
+		delete outPorts_[port];
+		outPorts_.erase(port);
+		u::log::print("[MP::closeOutPort] Port \"%s\" closed.\n",
+							port.c_str());
+		return 0;
+	}
+	return 2;
+	
+}
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+int closeInPort(std::string port){
+
+	if (port == ""){
+
+		// Create a list of open ports and destroy them all!
+		std::vector<std::string> ports_to_destroy;
+
+		for (auto const& [name, rtm] : inPorts_){
+			ports_to_destroy.push_back(name);
+		}
+
+		for (unsigned int i = 0; i < ports_to_destroy.size(); i++){
+			closeInPort(ports_to_destroy.at(i));
+		}
+		return 1;
+	}
+	// Let's check if this port is in the map //
+	if (inPorts_.count(port) > 0) {
+		// RtMidi port destructor closes connections on its own
+		delete inPorts_[port];
+		inPorts_.erase(port);
+		u::log::print("[MP::closeInPort] Port \"%s\" closed.\n",
+							port.c_str());
+		return 0;
+	}
+	return 2;
+	
+}
+
+/* -------------------------------------------------------------------------- */
+
+std::string getOutDeviceName(int index){
+	if (index < 0) return "";
+	try {return midiOut_->getPortName(index);}
+	catch (RtMidiError& error) {return "";}
```

Comment:
Do we need to return "" instead of throwing? Are there some real use cases?

Reply:
I think that's legacy code again, and if RtMidi complains for whatever reason it will simply return a "", which means "nope, nothing interesting in there", and that's what we care about.
Of course we might as well try and catch errors every time we dare to address RtMidi, but then again, there is not much we can do about its errors anyway. 
I plan to spend some time adding some logging, reporting and whatnot so the information about errors is not completely lost, but I think error handling in this case is reasonable.